File: Day3/square_root/sqrt_binary.py
def sqrt_binary(x, precision=1e-10):
    if x < 0:
        raise ValueError("Cannot compute square root of a negative number.")
    low, high = 0, max(1.0, x)
    while high - low > precision:
        mid = (low + high) / 2
        if mid * mid < x:
            low = mid
        else:
            high = mid
    return (low + high) / 2

# Example usage
print(sqrt_binary(9))  # Approximate value of sqrt(9)

File: Week1/Day3/square_root/sqrt_binary.c
#include <stdio.h>

double sqrt_binary_search(double x, double precision) {
    if (x < 0) {
        return -1; // Error: negative input
    }
    double low = 0, high = (x < 1) ? 1 : x;
    double mid;
    while ((high - low) > precision) {
        mid = (low + high) / 2;
        if (mid * mid < x) {
            low = mid;
        } else {
            high = mid;
        }
    }
    return (low + high) / 2;
}
int main() {
    double x = 9.0;
    double precision = 1e-10;
    printf("sqrt(%.2f) ≈ %.10f\n", x, sqrt_binary_search(x, precision));
     return 0;
}


